Update cmake find package review (#6691)
conan-io/conan-center-index: docs/reviewing.md

### Minimalistic Source Code

The contents of `test_package.c` or `test_package.cpp` should be as minimal as possible, including a few headers at most with simple instantiation of objects to ensure linkage
and dependencies are correct.

### Verifying Components

When components are defined in the `package_info` in `conanfile.py` the following conditions are desired

> * use `cmake_find_package` if library has an [official](cmake.org/cmake/help/latest/manual/cmake-modules.7.html#find-modules) CMake module emulated in the recipe.

Reviewer: What if a project installs a cmake script without targets or the documentation of CMake doesn't mention a target?
Is it okay to use cmake variables then? Eventually emulated with a included custom cmake module?

Reviewer 2: does it? there is no word `target` used, only word `module` (perhaps, meaning `find module`, as hyperlink suggests), which is way more generic, and includes both modules with variables and modules with targets.

Reviewer 3: The next sentence explicitly called out targets... just not in the UI diff

Reviewer 2: the next sentence is about `cmake_find_package_multi`, which is all about targets

> * use `cmake_find_package_multi` if library provides official cmake config file emulated in the recipe. If there are more than one target, try to use all of them, or add an other executable linking to the global (usually unofficial) target. There are some ways to identify when use it:

Reviewer 2: ```suggestion
* use `cmake_find_package_multi` if library provides an official cmake config file emulated in the recipe. If there are more than one target, try to use all of them, or add another executable linking to the global (usually unofficial) target. There are some ways to identify when to use it:
```

> * use `cmake_find_package_multi` if library provides an official cmake config file emulated in the recipe. If there are more than one target, try to use all of them, or add another executable linking to the global (usually unofficial) target. There are some ways to identify when to use it:

Reviewer 3: Like this?

```suggestion
* use `cmake_find_package_multi` if library provides an official cmake config file emulated in the recipe. If there are more than one target, try to use all of them, or add another executable linking to the global (usually unofficial) target. There may additionally be variables that are emulated by the recipe which should be used as well. There are some ways to identify when to use it:
```

Reviewer: Indeed, as long as the varaibles are mentioned.

>     * Usually, project install their cmake files on `package_folder/lib/cmake`. The folder are removed from package folder by calling `tools.rmdir(os.path.join(self.package_folder), "lib", "cmake")`

Reviewer 2: ```suggestion
    * Usually, project installs its cmake files into `package_folder/lib/cmake`. The folder is removed from package folder by calling `tools.rmdir(os.path.join(self.package_folder), "lib", "cmake")`
```

>     * Also, the upstream's cmakefile can use [install(EXPORT ..)](https://cmake.org/cmake/help/latest/command/install.html#export)

Reviewer 3: ```suggestion
    * Also, the library's CMake scripts can use [install(EXPORT ..)](https://cmake.org/cmake/help/latest/command/install.html#export) and/or may install [package configuration helpers](https://cmake.org/cmake/help/latest/module/CMakePackageConfigHelpers.html)
```

When components are defined in the `package_info` in `conanfile.py` the following conditions are desired

- use `cmake_find_package` if library has an [official](cmake.org/cmake/help/latest/manual/cmake-modules.7.html#find-modules) CMake module emulated in the recipe.

> - use `cmake_find_package_multi` if library provides official cmake config file emulated in the recipe. If there are more than one target, try to use all of

Reviewer 2: how do I know that?

Author: have to check the recipe directly. we already use the cmake_official_config.cmake. The official modules we can find here: https://cmake.org/cmake/help/latest/manual/cmake-modules.7.html#find-modules

Reviewer 3: https://cmake.org/cmake/help/latest/module/CMakePackageConfigHelpers.html

Is also a good indicator

Reviewer 2: @uilianries may you add some short instructions for newbiews, what should be done and how to check if library provides an official cmake config file? it's not obvious right now.

Reviewer 4: It's not a silver bullet, but when you have to remove `cmake` folder in `lib` or `share` folder, library likely provides an official CMake config file.

Reviewer 2: there might be some rules, e.g. if recipe contains any of:
```
tools.rmdir(os.path.join(self.package_folder), "lib", "cmake")
tools.rmdir(os.path.join(self.package_folder), "share", "cmake")
```
it should use `cmake_find_package[_multi]` generator, from your words, it sounds like a regularity.
another possible rule, if recipe contains any of:
```
.names["cmake_find_package"]
.names["cmake_find_package_multi"]
```
it should use `cmake_find_package[_multi]` generator.
is it accurate?

Reviewer 4: yes, and I prefer the last rule.

there are no `tools.rmdir(os.path.join(self.package_folder), "share", "cmake")` in recipes, because we remove `share` folder directly.

Reviewer 2: @uilianries I also can't find any information on what is the `cmake_official_config.cmake`. can you explain that one as well?

Reviewer 4: precedence:
- `self.cpp_info.filenames["cmake_find_package_multi"]`
- `self.cpp_info.names["cmake_find_package_multi"]`
- recipe name

Reviewer 2: @SpaceIm these might be set on components as well:
```
        self.cpp_info.components["headers"].names["cmake_find_package"] = "headers"
        self.cpp_info.components["headers"].names["cmake_find_package_multi"] = "headers"
```
global names/filenames aren't mandatory, if library name matches cmake name (?)

Reviewer 4: yes, I though that you were talking about config file name

Anyway, I strongly advice to be explicit and not rely on recipe name inheritence.

Reviewer 5: There is also `set_property("cmake_target_name", ...)` and `set_property("cmake_file_name", ...)` now

Reviewer 4: But only for `CMakeDeps`?

Reviewer 2: nope, old generators also can use `set_property` (they are wrapped for comaptibility).
so the full list of patterns hook may scan for:

- `.names["cmake_find_package"]`
- `.names["cmake_find_package_multi"]`
- `.filenames["cmake_find_package"]`
- `.filenames["cmake_find_package_multi"]`
- `set_property("cmake_target_name"`
- `set_property("cmake_file_name"`
- `tools.rmdir(os.path.join(self.package_folder), "lib", "cmake")`
- `tools.rmdir(os.path.join(self.package_folder), "share", "cmake")` (unlikely to be encountered in the wild)

recipe name, I an not sure if should be checked at all. this list is dynamic and changes from one version of CMake to another.
